src/zarr/codecs/sharding.py

```diff
@@ -760,6 +771,115 @@ async def _load_full_shard_maybe(
             else None
         )
 
+    async def _load_partial_shard_maybe(
+        self,
+        byte_getter: ByteGetter,
+        prototype: BufferPrototype,
+        chunks_per_shard: ChunkCoords,
+        all_chunk_coords: set[ChunkCoords],
+    ) -> ShardMapping | None:
+        """
+        Read chunks from `byte_getter` for the case where the read is less than a full shard.
+        Returns a mapping of chunk coordinates to bytes or None.
+        """
+        shard_index = await self._load_shard_index_maybe(byte_getter, chunks_per_shard)
```

Comment:
this will fetch the shard index every time. Should we push reading the shard index higher up in the stack, and have this function take the content of the index as a parameter? This might be out of scope for this PR.

Reply:
(we could also build caching into the `_load_shard_index` function)